File: tests/capture-screenshots.js
/**
 * TTS MultiModel - Full Website Screenshot Capture
 *
 * 由 Seedvr2 / Image_MultiModel 的 tests/capture-screenshots.js 复制改造：
 *   - BASE_URL 改为 http://127.0.0.1:7869（TTS MultiModel 默认端口）
 *   - 主题持久化键改为 app_theme（class 方式，非 data-theme）
 *   - 页面为单页 SPA + htmx 标签页（sidebar 的 data-tab 切换 /tab/* 到 #tab-content）
 *   - 健康检查端点改为 /api/system/health
 *   - 新增：自动关闭新手引导弹窗（tts_onboarded_v1）
 *
 * Prerequisites:
 *   - TTS MultiModel server running (default http://127.0.0.1:7869), start with start.bat
 *   - Playwright chromium installed: npm install && npx playwright install chromium
 *
 * Usage:
 *   node capture-screenshots.js
 *
 * Optional env overrides:
 *   TTS_BASE_URL  e.g. http://127.0.0.1:7869
 *   TTS_OUT_DIR   e.g. ./screenshots
 *
 * Output: screenshots/<viewport>/<theme>/<NN>-<name>.png
 *
 * NOTE: 只通过 JS click 切换纯 UI 状态（主题、标签页）。触发真实后端工作的
 *       （生成、克隆、训练等）按钮不点击。
 */
const { chromium } = require('@playwright/test');
const path = require('path');
const fs = require('fs');

const BASE_URL = process.env.TTS_BASE_URL || 'http://127.0.0.1:7869';
const OUTPUT_DIR = process.env.TTS_OUT_DIR
  ? path.resolve(process.env.TTS_OUT_DIR)
  : path.join(__dirname, '..', 'screenshots');

const VIEWPORTS = {
  desktop: { width: 1920, height: 1080 },
  tablet: { width: 768, height: 1024, isMobile: true, hasTouch: true },
  mobile: { width: 375, height: 812, isMobile: true, hasTouch: true },
};

const THEMES = ['dark', 'light'];

// 与 tests/e2e/test_screenshot_capture.py 的 TABS 保持一致的核心标签页
const TABS = [
  { num: '01', tab: 'voice_design' },
  { num: '02', tab: 'voice_clone' },
  { num: '03', tab: 'ultimate_clone' },
  { num: '04', tab: 'script' },
  { num: '05', tab: 'prompt_continue' },
  { num: '06', tab: 'lora' },
  { num: '07', tab: 'lora_training' },
  { num: '08', tab: 'settings' },
  { num: '09', tab: 'history' },
  { num: '10', tab: 'persona' },
  { num: '11', tab: 'help' },
];

function ensureDir(dir) {
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
}

async function screenshotPage(page, name, options = {}) {
  const {
    fullPage = true,
    waitFor = null,
    viewportName = 'desktop',
    theme = 'dark',
  } = options;

  const dir = path.join(OUTPUT_DIR, viewportName, theme);
  ensureDir(dir);

  const filePath = path.join(dir, `${name}.png`);

  if (waitFor) {
    await page.waitForTimeout(waitFor);
  }

  await page.screenshot({ path: filePath, fullPage });
  console.log(`  Captured: ${filePath}`);
}

async function setTheme(page, theme) {
  // 主题持久化键 'app_theme' 是 static/js/theme_lang.js 使用的真实键。
  // 通过 class（dark/light）+ colorScheme 应用，导航前设置 localStorage 即可。
  await page.evaluate((t) => {
    localStorage.setItem('app_theme', t);
    const html = document.documentElement;
    html.classList.remove('light', 'dark');
    html.classList.add(t);
    html.style.colorScheme = t;
  }, theme);
  await page.waitForTimeout(300);
}

async function safe(label, viewportName, theme, fn) {
  try {
    await fn();
  } catch (e) {
    console.error(`  [SKIP] ${label} (${viewportName}, ${theme}): ${e.message}`);
  }
}

// 通过 JS click 切换标签页（htmx 处理 /tab/* 加载），避免元素遮挡或移动端侧栏隐藏
async function clickTab(page, tabName) {
  await page.evaluate((t) => {
    const el = document.querySelector(`.sidebar-item[data-tab="${t}"]`);
    if (!el) throw new Error(`sidebar item [data-tab=${t}] not found`);
    el.click();
  }, tabName);
}

async function captureHomePage(page, viewportName, theme) {
  console.log(`Capturing Home Page (${viewportName}, ${theme})...`);
  await page.goto(`${BASE_URL}/`, { waitUntil: 'domcontentloaded', timeout: 30000 });
  await page.waitForTimeout(3000);

  await screenshotPage(page, '00-home', { viewportName, theme });
}

async function captureTab(page, tab, viewportName, theme) {
  const label = `${tab.num}-${tab.tab}`;
  console.log(`Capturing Tab ${label} (${viewportName}, ${theme})...`);

  await safe(label, viewportName, theme, async () => {
    await page.goto(`${BASE_URL}/`, { waitUntil: 'domcontentloaded', timeout: 30000 });
    await page.waitForTimeout(1500);
    await clickTab(page, tab.tab);
    // htmx 加载 + 渲染标签页内容
    await page.waitForTimeout(1800);
    await page.waitForSelector('#tab-content', { timeout: 5000 });
    await screenshotPage(page, `${tab.num}-${tab.tab}`, { viewportName, theme });
  });
}

async function captureAllViewports(page, viewports, themes) {
  for (const [vpName, vpSize] of Object.entries(viewports)) {
    console.log(`\n=== Viewport: ${vpName} (${vpSize.width}x${vpSize.height}) ===`);
    await page.setViewportSize({ width: vpSize.width, height: vpSize.height });

    for (const theme of themes) {
      console.log(`\n--- Theme: ${theme} ---`);
      await setTheme(page, theme);

      await captureHomePage(page, vpName, theme);
      for (const tab of TABS) {
        await captureTab(page, tab, vpName, theme);
      }
    }
  }
}

(async () => {
  console.log('TTS MultiModel - Full Website Screenshot Capture');
  console.log('=================================================');
  console.log(`Base URL: ${BASE_URL}`);
  console.log(`Output Dir: ${OUTPUT_DIR}`);
  console.log('');

  ensureDir(OUTPUT_DIR);

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext();
  const page = await context.newPage();

  // 拦截外部字体/CDN 请求（如 fonts.googleapis.com / fonts.gstatic.com）。
  // base.html 头部的 Google Fonts 样式表是 render-blocking 的，在无外网环境下会
  // 挂起并导致 domcontentloaded 永不触发；本地资源（127.0.0.1）照常放行。
  await context.route('**/*', (route) => {
    const url = route.request().url();
    if (url.startsWith('http://fonts.') || url.startsWith('https://fonts.')) {
      return route.abort();
    }
    return route.continue();
  });

  // 自动关闭 TTS 新手引导弹窗（避免遮挡截图）
  await page.addInitScript(() => {
    localStorage.setItem('tts_onboarded_v1', 'true');
  });

  try {
    console.log('Checking if server is running...');
    try {
      await page.goto(`${BASE_URL}/api/system/health`, { timeout: 10000 });
      console.log('Server is running!');
    } catch (e) {
      console.error('ERROR: Server is not running at', BASE_URL);
      console.error('Please start the server first with: start.bat');
      process.exit(1);
    }

    await captureAllViewports(page, VIEWPORTS, THEMES);

    console.log('\n=========================================');
    console.log('Screenshot capture complete!');
    console.log(`All screenshots saved to: ${OUTPUT_DIR}`);

  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  } finally {
    await browser.close();
  }
})();

File: scripts/capture_screenshots.py
"""统一截图脚本 — Python Playwright 版（替代 Node.js 版）。

功能与 tests/capture-screenshots.js 完全一致，但使用 Python Playwright API，
消除 Node.js 依赖，统一技术栈。

用法::

    # 1. 启动服务器
    python bin/start_ui_test.py

    # 2. 运行截图脚本
    python scripts/capture_screenshots.py

    # 或指定 URL
    python scripts/capture_screenshots.py --url http://127.0.0.1:7869

输出目录: docs/screenshots/
"""

from __future__ import annotations

import argparse
import os
import sys
import time
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
OUTPUT_DIR = REPO_ROOT / "docs" / "screenshots"
DARK_DIR = OUTPUT_DIR / "dark"

VIEWPORTS = {
    "desktop": {"width": 1366, "height": 900},
}

TABS = [
    {"num": "01", "tab": "voice_design", "name": "voice_design"},
    {"num": "02", "tab": "voice_clone", "name": "voice_clone"},
    {"num": "03", "tab": "ultimate_clone", "name": "ultimate_clone"},
    {"num": "04", "tab": "script", "name": "script_workshop"},
    {"num": "05", "tab": "prompt_continue", "name": "prompt_continuation"},
    {"num": "06", "tab": "lora", "name": "lora"},
    {"num": "07", "tab": "lora_training", "name": "lora_training"},
    {"num": "08", "tab": "settings", "name": "settings"},
    {"num": "09", "tab": "history", "name": "history"},
    {"num": "10", "tab": "persona", "name": "persona_library"},
    {"num": "11", "tab": "help", "name": "help"},
]


def ensure_dir(path: Path) -> None:
    path.mkdir(parents=True, exist_ok=True)


def set_theme(page, theme: str) -> None:
    """设置页面主题（light/dark）。"""
    page.evaluate(f"""
        () => {{
            localStorage.setItem('tts_onboarded_v1', '1');
            localStorage.setItem('app_theme', '{theme}');
            document.documentElement.classList.remove('dark', 'light');
            document.documentElement.classList.add('{theme}');
            document.documentElement.style.colorScheme = '{theme}';
        }}
    """)
    page.wait_for_timeout(300)


def click_tab(page, tab_name: str) -> bool:
    """点击侧边栏标签页按钮。"""
    button = page.locator(f".sidebar-item[data-tab='{tab_name}']")
    if button.count() == 0:
        print(f"  [skip] Tab button not found: data-tab='{tab_name}'")
        return False
    button.first.click()
    page.wait_for_load_state("networkidle", timeout=15000)
    page.wait_for_function(
        "() => { const el = document.querySelector('#tab-content'); "
        "return el && el.innerHTML.trim().length > 0; }",
        timeout=5000,
    )
    return True


def capture_tab(page, tab: dict, theme: str) -> bool:
    """捕获单个标签页截图。"""
    suffix = "_dark_viewport" if theme == "dark" else "_viewport"
    filename = f"voxcpm2_{tab['num']}_{tab['name']}{suffix}.png"
    out_dir = DARK_DIR if theme == "dark" else OUTPUT_DIR
    file_path = out_dir / filename

    page.goto("http://127.0.0.1:7869/", wait_until="domcontentloaded", timeout=30000)
    page.wait_for_load_state("networkidle", timeout=15000)
    page.wait_for_timeout(800)
    set_theme(page, theme)

    if not click_tab(page, tab["tab"]):
        return False

    page.screenshot(path=str(file_path), full_page=True)
    rel_path = file_path.relative_to(REPO_ROOT)
    print(f"  Captured: {rel_path}")
    return True


def capture_home_page(page, theme: str) -> None:
    """捕获首页截图。"""
    suffix = "_dark_viewport" if theme == "dark" else "_viewport"
    filename = f"home{suffix}.png"
    out_dir = DARK_DIR if theme == "dark" else OUTPUT_DIR
    file_path = out_dir / filename

    page.goto("http://127.0.0.1:7869/", wait_until="domcontentloaded", timeout=30000)
    page.wait_for_load_state("networkidle", timeout=15000)
    page.wait_for_timeout(2000)
    set_theme(page, theme)
    page.screenshot(path=str(file_path), full_page=True)
    rel_path = file_path.relative_to(REPO_ROOT)
    print(f"  Captured: {rel_path}")


def main():
    parser = argparse.ArgumentParser(description="TTS MultiModel Screenshot Capture (Python)")
    parser.add_argument("--url", default="http://127.0.0.1:7869", help="Server URL")
    args = parser.parse_args()

    try:
        from playwright.sync_api import sync_playwright
    except ImportError:
        print("ERROR: Playwright not installed. Install with: pip install playwright && playwright install")
        sys.exit(1)

    print("TTS MultiModel - Screenshot Capture (Python)")
    print("=" * 44)
    print(f"Base URL: {args.url}")
    print(f"Output:   {OUTPUT_DIR}")
    print(f"Dark sub: {DARK_DIR}")
    print()

    ensure_dir(OUTPUT_DIR)
    ensure_dir(DARK_DIR)

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context()
        page = context.new_page()

        print("Checking if server is running...")
        try:
            resp = page.goto(args.url, timeout=10000, wait_until="domcontentloaded")
            if not resp or not resp.ok:
                raise ConnectionError(f"server responded {resp.status if resp else 'no response'}")
            print("Server is running!")
        except Exception as e:
            print(f"ERROR: Server is not running at {args.url}")
            print(f"  {e}")
            sys.exit(1)

        for vp_name, vp_size in VIEWPORTS.items():
            print(f"\n=== Viewport: {vp_name} ({vp_size['width']}x{vp_size['height']}) ===")
            page.set_viewport_size(vp_size)

            for theme in ["light", "dark"]:
                print(f"\n--- Theme: {theme} ---")
                try:
                    capture_home_page(page, theme)
                except Exception as e:
                    print(f"  home failed: {e}")

                for tab in TABS:
                    try:
                        capture_tab(page, tab, theme)
                    except Exception as e:
                        print(f"  [err] {tab['tab']}: {e}")

        print("\n" + "=" * 44)
        print("Screenshot capture complete!")
        browser.close()


if __name__ == "__main__":
    main()
